third_party/bigframes_vendored/ibis/expr/operations/analytic.py

```diff
@@ -2,22 +2,24 @@
 
 from __future__ import annotations
 
-from ibis.expr.operations.analytic import Analytic
+import ibis.expr.datatypes as dt
+import ibis.expr.operations.analytic as ibis_ops_analytic
+import ibis.expr.operations.core as ibis_ops_core
 import ibis.expr.rules as rlz
 
 
-class FirstNonNullValue(Analytic):
+class FirstNonNullValue(ibis_ops_analytic.Analytic):
     """Retrieve the first element."""
 
-    arg = rlz.column(rlz.any)
-    output_dtype = rlz.dtype_like("arg")
+    arg: ibis_ops_core.Column[dt.Any]
+    dtype = rlz.dtype_like("arg")
 
 
-class LastNonNullValue(Analytic):
+class LastNonNullValue(ibis_ops_analytic.Analytic):
```

Comment:
Since we have been breaking several internal parts of the codebase (sorry for that), maintaining custom operations this way is probably inconvenient. I wanted to have a documented and stable way of creating custom nodes for a long time now, just didn't have enough usage examples which we could draft a desired API from. This might be another topic we should discuss in our issue tracker.

Reply:
Good call. Filed https://github.com/ibis-project/ibis/issues/7767

I don't even know that it needs to be a stable interface for custom ops. Maybe just an escape hatch of some sort for raw SQL that's just a Value, not a full table expression like the current Backend.sql() escape hatch. That said, such an escape hatch is going to look a lot like a custom op, since ideally we'd propagate type info somehow so that expressions can continue to be built on it.